packages/database/src/server/changeHandlers/entityHierarchyCacher/EntityParentChildRelationBuilder.js

```diff
@@ -0,0 +1,226 @@
+import { ORG_UNIT_ENTITY_TYPES } from '../../../core/modelClasses/Entity';
+
+/**
+ * Builds and caches the entity parent child relations for a given hierarchy
+ * This never wipes the subtrees and rebuilds, always only adds new ones and deletes the obsolete ones
+ * in order to keep the data to be synced minimal
+ */
+export class EntityParentChildRelationBuilder {
+  constructor(models) {
+    this.models = models;
+  }
+
+  /**
+   * Rebuilds the entity parent child relations for the given rebuild jobs
+   * @param {{hierarchyId: string; rootEntityId: string}[]} rebuildJobs
+   */
+  async rebuildRelations(rebuildJobs) {
+    // projects are the root entities of every full tree, so start with them
+    for (const { hierarchyId, rootEntityId } of rebuildJobs) {
+      const project = await this.models.project.findOne({ entity_hierarchy_id: hierarchyId });
+      await this.rebuildRelationsForEntity(hierarchyId, rootEntityId, project);
+    }
+  }
+
+  /**
+   * Rebuilds the entity parent child relations for a given project
+   * @param {import('../../../core/modelClasses/Project').Project} project
+   */
+  async rebuildRelationsForProject(project) {
+    const { entity_id: projectEntityId, entity_hierarchy_id: hierarchyId } = project;
+    await this.rebuildRelationsForEntity(hierarchyId, projectEntityId, project);
+  }
+
+  /**
+   * Rebuilds the entity parent child relations for a given entity
+   * It traverses the hierarchy from the root entity to the leaves, and caches new relations + deletes obsolete ones
+   * @param {string} hierarchyId
+   * @param {string} rootEntityId
+   * @param {import('../../../core/modelClasses/Project').Project} project
+   */
+  async rebuildRelationsForEntity(hierarchyId, rootEntityId, project) {
+    const { entity_id: projectEntityId } = project;
+    await this.fetchAndCacheChildren(hierarchyId, [rootEntityId]);
+    await this.deleteOrphanedRelations(hierarchyId, projectEntityId);
+  }
+
+  async fetchAndCacheChildren(hierarchyId, parentIds, childrenAlreadyCached = new Set()) {
+    const entityRelationChildCount = await this.countEntityRelationChildren(hierarchyId, parentIds);
+    const useEntityRelationLinks = entityRelationChildCount > 0;
+
+    // Generate the new relations for this level
+    const validParentChildIdPairs = useEntityRelationLinks
```

Comment:
Not your code, but `use` prefix feels a _touch_ odd?

```suggestion
    const hasEntityRelationChildren = entityRelationChildCount > 0;

    // Generate the new relations for this level
    const validParentChildIdPairs = hasEntityRelationChildren
```

…or could even inline it